### PR TITLE
[ENH] RESS implem sklearn interface

### PR DESCRIPTION
**Pull request by ludovicdmt**
Hello, 

I was recently asked to adapt your RESS implementation so it can be compatible with:
* MNE formatting `(n_trials, n_chans, n_samples)` (instead of `(n_samples, n_chans, n_trials)` in the current implem)
* Sklearn interface: class inherited from `BaseEstimator` and `TransformerMixin` instead of function and `fit` and `transform` methods

I basically used your implem and just simplified some parts of it: removed the `fold`/`unfold`/`demean` function to use vectorization from Numpy, use sklearn to compute empirical covariance matrices (can be easily adapted to use another covariance matrix estimation method with regul).
I compared the two implementations on randomly generated data and they outputted the same results (max difference about $10${-16}$). 

I also have updated accordingly the `test_ress.py` so it can run with the new data formatting and the transformation of RESS to a class.

### CODE REVIEW

File: meegkit/ress.py [hunk 1]
```diff
@@ -29,92 +26,236 @@ def RESS(X, sfreq: int, peak_freq: float, neig_freq: float = 1,
...
+    def _avg_cov(self, X):
+        """Compute average covariance matrix along trials"""
+        X_transpose = X.transpose([0, 2, 1])
+        s = list(X_transpose.shape)
+        combined = functools.reduce(lambda x, y: x * y, s[0: 0 + 1 + 1])
+        X_combined = np.reshape(X_transpose, s[:0] + [combined] + s[0 + 1 + 1:])
+        cov = EmpiricalCovariance().fit(X_combined)
+        return cov.covariance_
+
+    def _gaussfilt(
```
Comment (nbara) [inline meegkit/ress.py:187]: I have other projects that use the gaussfilt function (not related to RESS at all). Can you not keep this in a separate function?

File: meegkit/ress.py [hunk 2]
```diff
@@ -29,92 +26,236 @@ def RESS(X, sfreq: int, peak_freq: float, neig_freq: float = 1,
...
-    >>> # Then matrix multiply each trial by the unmixing matrix:
-    >>> proj = matmul3d(out, fromRESS)
-
-    To transform a new observation into RESS component space (e.g. in the
-    context of a cross-validation, with separate train/test sets):
-
-    >>> # Start by applying RESS to the train set:
-    >>> out, _, toRESS = ress.RESS(data, sfreq, peak_freq, return_maps=True)
-    >>> # Then multiply your test data by the toRESS:
-    >>> new_comp = new_data @ toRESS
```
Comment (nbara) [inline meegkit/ress.py:63]: I'd like to keep this bit of documentation if possible

File: meegkit/ress.py [hunk 3]
```diff
@@ -29,92 +26,236 @@ def RESS(X, sfreq: int, peak_freq: float, neig_freq: float = 1,
         FWHM of the neighboring frequencies (default=1).
     n_keep : int
         Number of components to keep (default=1). -1 keeps all components.
-    return_maps : bool
```
Comment (nbara) [inline meegkit/ress.py:32]: This should become a class attribute I think

**Review by nbara — CHANGES_REQUESTED** [review]

**Comment by nbara** [comment]
> Hello,
> 
> I was recently asked to adapt your RESS implementation so it can be compatible with:
> 
> * MNE formatting `(n_trials, n_chans, n_samples)` (instead of `(n_samples, n_chans, n_trials)` in the current implem)

I'm a bit ambivalent about this TBH. The convention for most functions in meegkit is `(n_samples, n_chans, n_trials)`. 

It's true that MNE compatibility is nice, but this is a breaking change that would likely impact existing code for all users of the toolbox...

> * Sklearn interface: class inherited from `BaseEstimator` and `TransformerMixin` instead of function and `fit` and `transform` methods
> 
> I basically used your implem and just simplified some parts of it: removed the `fold`/`unfold`/`demean` function to use vectorization from Numpy, use sklearn to compute empirical covariance matrices (can be easily adapted to use another covariance matrix estimation method with regul). I compared the two implementations on randomly generated data and they outputted the same results (max difference about 10{-16}$).

👍👍 

> I also have updated accordingly the `test_ress.py` so it can run with the new data formatting and the transformation of RESS to a class.

👍

File: meegkit/ress.py (hunk 3 above)
Comment (ludovicdmt) [reply meegkit/ress.py:32]: Then the name should change. In the original implem no matter `return_maps` the mixing and unmixing matrices are computed but returned only if `True`. 

In this implem both mixing and unmixing are also computed during the `fit` (`fit` returns `self`). Mixing matrix is mandatory for the transform so it has to be computed but unmixing matrix could be optional. 

Then I could add `compute_unmixing` attribute.

File: meegkit/ress.py (hunk 1 above)
Comment (ludovicdmt) [reply meegkit/ress.py:187]: BTW I thing there is something strange like a redundant copy-paste in `gaussifilt` function (line 338) but maybe this is correct ?

**Comment by nbara** [comment]
HI @ludovicdmt there are still conflicts in `ress.py` apparently. 

>Here in ISAE (France, Toulouse) we use exclusively MNE for data loading/processing so this motivated the change but it is probably different elsewhere. 

MNE-style ordering (trials first) makes sense for mostly offline analyses, but sample first makes more sense when a) there are no trials (it's much easier to consider the last dimension as optional) b) for real-time applications when the data needs to be serialised.

**Comment by ludovicdmt** [comment]
Yep you are right in the buffer for my real time applications, I have data in 2D format. 
Though I artificially add a dimension (simulating single trial) to still be able to use some convenient MNE functions but that's probably not the most efficient way...

**Review by nbara — APPROVED** [review]

**Comment by nbara** [comment]
Hey @ludovicdmt finally found some time to look into this. 

I reverted the MNE ordering, but kept the sklearn style interface

Merging this, thanks!

**Pull request merged**